src/layout/layout.rs

```diff
@@ -42,6 +41,22 @@ thread_local! {
     ));
 }
 
+// Similar to `itertools::Itertools::tuple_combinations`.
+fn tuple_combinations<T, E>(
+    slice: &[T],
+    mut f: impl FnMut((&T, &T)) -> Result<(), E>,
+) -> Result<(), E> {
+    if !slice.is_empty() {
+        for (ind, left) in slice[..slice.len() - 1].iter().enumerate() {
+            for right in &slice[ind + 1..] {
+                f((left, right))?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
```

Comment:
I'd prefer not to have small utils that are defined in itertools like this as it indicates that we have some sort of special reason that we're doing this differently, which is an extra cognitive burden when maintaining the crate.